src/ser/mod.rs

```diff
@@ -276,8 +276,12 @@ impl<'a> ser::Serializer for &'a mut Serializer {
                     self.buf.push(hex2);
                 }
                 _ => {
-                    let encoded = c.encode_utf8(&mut encoding_tmp as &mut [u8]);
-                    self.buf.extend_from_slice(encoded.as_bytes());
+                    if c.len_utf8() == 1 {
```

Comment:
Nice one. We can move this to the switch and avoid the call to len_utf8. This us just:

```
#[inline]
fn len_utf8(code: u32) -> usize {
    if code < MAX_ONE_B {
        1
    } else if code < MAX_TWO_B {
        2
    } else if code < MAX_THREE_B {
        3
    } else {
        4
    }
}
```

with `const MAX_ONE_B: u32 = 0x80;`. See also https://stackoverflow.com/a/33349765

Reply:
Good idea. Extracting the last bits of juice. :-)

I'll try, but with `len_utf8` already being `inline`, it probably won't make much of a difference.

Reply:
I did it, and it's practically the same. Leaving as it is now for clarity.

Reply:
Good, thanks for checking